[dev-1.30] KEP-3857: Recursive Read-only (RRO) mounts (#45159)
kubernetes/website: content/en/docs/concepts/storage/volumes.md

  (unmounted) by the containers on termination.
  {{< /warning >}}

## Read-only mounts

A mount can be made read-only by setting the `.spec.containers[].volumeMounts[].readOnly`

> field to `true`.

Reviewer: ```suggestion
field to `true`.  This does not make the volume itself read-only, but that specific
container will not be able to write to it.  Other containers in the Pod may mount
the same volume as read-write.
```

Author: done

On Linux, read-only mounts are not recursively read-only by default.

> For example, if you have another mount such as tmpfs, NFS, or USB storage
> file system under `/mnt/<SUBMOUNT>`, even if you specify to mount `/mnt` into a
> container as read-only, `/mnt/<SUBMOUNT>` will become still writable.

Reviewer: ```suggestion
For example, consider a Pod which mounts the hosts `/mnt` as a `hostPath` volume.  If
there is another filesystem mounted read-write on `/mnt/<SUBMOUNT>` (such as tmpfs,
NFS, or USB storage), the volume mounted into the container(s) will also have a writeable
`/mnt/<SUBMOUNT>`, even if the mount itself was specified as read-only.
```

Author: done

### Recursive read-only mounts

{{< feature-state feature_gate_name="RecursiveReadOnlyMounts" >}}

Recursive read-only mounts can be enabled by setting the
`RecursiveReadOnlyMounts` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
for kubelet and kube-apiserver, and setting the `.spec.containers[].volumeMounts[].recursiveReadOnly`
field for a pod.

The allowed values are:

* `Disabled` (default): no effect.

* `Enabled`: makes the mount recursively read-only.
  Needs all the following requirements to be satisfied:
  * `readOnly` is set to `true`
  * `mountPropagation` is unset, or, set to `None`
  * The host is running with Linux kernel v5.12 or later

>   * The CRI-level container runtime supports recursive read-only mounts

Author: done

>   * The OCI-level container runtime supports recursive read-only mounts.

Reviewer: Comment that it will fail if any of these is not true?

Author: done

  (unmounted) by the containers on termination.
  {{< /warning >}}

## Read-only mounts

A mount can be made read-only by setting the `.spec.containers[].volumeMounts[].readOnly`
field to `true`.

On Linux, read-only mounts are not recursively read-only by default.
For example, if you have another mount such as tmpfs, NFS, or USB storage
file system under `/mnt/<SUBMOUNT>`, even if you specify to mount `/mnt` into a
container as read-only, `/mnt/<SUBMOUNT>` will become still writable.

### Recursive read-only mounts

> {{< feature-state for_k8s_version="v1.30" state="alpha" >}}

Reviewer 2: (nit / even better)
```suggestion
{{< feature-state feature_gate_name="RecursiveReadOnlyMounts" >}}
```

Author: done

Recursive read-only mounts can be enabled by setting the
`RecursiveReadOnlyMounts` [feature gate](/docs/reference/command-line-tools-reference/feature-gates/)
for kubelet and kube-apiserver, and setting the `.spec.containers[].volumeMounts[].recursiveReadOnly`
field for a pod.

The allowed values are:

* `Disabled` (default): no effect.

* `Enabled`: makes the mount recursively read-only.
  Needs all the following requirements to be satisfied:
  * `readOnly` is set to `true`
  * `mountPropagation` is unset, or, set to `None`
  * The host is running with Linux kernel v5.12 or later

>   * The runtime class supports recursive read-only mounts.
>     e.g., runc v1.1 or later.

Reviewer 2: ```suggestion
  * The OCI-level container runtimesupports recursive read-only mounts.
```
I've dropped a line because:
- we [avoid Latin phrases](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-latin-phrases)
- we should link to a page about `runc`, but then we also have a [marker](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#third-party-content-item) to add when we do that.

Author: done

* `IfPossible`: attempts to apply `Enabled`, and falls back to `Disabled`
  if the feature is not supported by the kernel or the runtime class.

> Example:

Reviewer 2: (nit)

Ideally (and we'll insist on this for beta): use a [downloadable code sample](https://kubernetes.io/docs/contribute/style/hugo-shortcodes/#source-code-files). We also test that these samples validate, whereas we don't do that for YAML snippets in pages.

Author: done
